### ISSUE TITLE
kubectl delete --all: does not include uninitialized resources

### ISSUE THREAD

**Issue by ahmetb** [kubernetes/kubernetes#49035]
tl;dr Similar to #49032, `kubectl delete --all` command does not delete the uninitialized resources. It appears like it does not query the resource list with  `?includeUninitialized=true` in the first place.

Combined with #49032, these two issues prevents users from seeing or deleting uninitialized resources.

### Update: Proposed solution in comments

See https://github.com/kubernetes/kubernetes/issues/49035#issuecomment-320809074 for the proposed course of action.

## Repro steps

- Have kubectl 1.7 cluster with --runtime-config=admissionregistration.k8s.io/v1alpha1 enabled.
- Apply the initializer configuration below

      apiVersion: admissionregistration.k8s.io/v1alpha1
      kind: InitializerConfiguration
      metadata:
      name: example-config
      initializers:
      - name: podimage.example.com
          rules:
          - apiGroups:
              - ""
              apiVersions:
              - v1
              resources:
              - pods

- Run some pods: `kubectl run web --image=nginx --port 80 --replicas=5`
    - `kubectl get pods` won't show anything due to #49032
- Make API call with client-go listing pods (use `metav1.ListOptions{IncludeUninitialized: true}` and observe pods actually exist
- `kubectl delete pods --all` **doesn't delete anything** (prints "No resources found"). Running with -v=10 reveals:
    ```
    I0717 10:30:16.470350   76682 round_trippers.go:405] GET https://35.184.184.151/api/v1/namespaces/default/pods 200 OK in 40 milliseconds
    I0717 10:30:16.470373   76682 round_trippers.go:411] Response Headers:
    I0717 10:30:16.470378   76682 round_trippers.go:414]     Content-Type: application/json
    I0717 10:30:16.470382   76682 round_trippers.go:414]     Content-Length: 131
    I0717 10:30:16.470386   76682 round_trippers.go:414]     Date: Mon, 17 Jul 2017 17:30:16 GMT
    I0717 10:30:16.470413   76682 request.go:991] Response Body: {"kind":"PodList","apiVersion":"v1","metadata":{"selfLink":"/api/v1/namespaces/default/pods","resourceVersion":"4801"},"items":[]}
    ```

- `kubectl delete pod NAME` works (given you found out the names by querying the API directly with client-go)

------
kubectl version: 1.7.1

/sig cli
/area bug

**Comment by caesarxuchao** [comment]
We can introduce a global flag --include-uninitialized (default to "false") to kubectl. It's useful for debugging purpose as well, users should be able to use `kubectl get --include-uninitialized` to list unitialized objects.

Any kubectl commands that involve GET will need this flag, e.g., `kubectl edit`.

**Comment by ahmetb** [comment]
My main feeling is that all commands except `kubectl get`, `kubectl delete` should have an **implicit** `--include-initialized=true`.

It doesn't make sense to me that `kubectl scale`,  `kubectl edit` or `kubectl annotate` would fail to find the resource because it is not initialized yet. The resource is there, it can be edited/annotated etc. It just has not been fully initialized. 

Initializers are an implementation detail to most people, especially to _users_ of a cluster, who are not configuring the Initializers or InitializerConfigurations. If all these commands don't have an implicit `--include-initialized=true`, it will just confuse people. Perhaps even `kubectl get` should have `--include-initialized=true` by default. 

For `kubectl delete`: I would prefer if `--all` worked just fine. `kubectl delete pods --all --include-uninitialized` sounds a bit unnecessary. When I mean `--all`, I mean "_select all resources in the namespace of the specified resource types_" as the --help indicates.

**Comment by smarterclayton** [comment]
Before we add the flag, we should ask what users expect.  I think users expect `kubectl delete pods --all` to delete uninitialized pods.

**Comment by caesarxuchao** [comment]
THe apiserver hides uninitialized objects by default from LIST and WATCH operations, unless `ListOptions.IncludeUninitialized=true`, so @droot, @mengqiy and I went through all the kubectl commands that involve the `LIST` or `WATCH` operation:
* `kubectl delete <resource> --all`, `kubectl get <resource> --show-all`, and `kubectl apply --prune`: almost certainly they should include uninitialized objects by default.
* `kubectl get <resource>` and `kubectl describe <resource>`: I lean toward let them include uninitialized objects by default. Because otherwise it leads to confusion in theses cases:
  * apiserver responses 409 conflict to a create request while `kubectl get <resource>` doesn't include an object of that name, because it's uninitialized.
  * `kubectl get <resource> name` shows the object while `kubectl get <resource>` doesn't include the object.
* `kubectl get <resouce-name> -w`: should be consistent with `kbuectl get <resource>`
* `kubectl set`, `kubectl edit`, `kubectl label`, and `kubectl annotate`: these commands mutate the objects. I think by default they shouldn't include the uninitialized objects. We should discourage users from mutating uninitialized objects. @smarterclayton do you agree?

We plan to introduce `--include-uninitialized` as a global flag to kubectl, because it affects quite a few commands.

We also need to update the printers and describers to print out if an object is initialized.

cc @pwittrock

**Comment by smarterclayton** [comment]
Re label and annotate, what if someone wants to write an initializer in
bash that runs those commands?  Definitely needs to be possible.  I think
we probably need to think of all mutating commands as "how would the user
react to being denied edit"?  I think omitting by default makes sense for
scripting commands - i.e. kubectl annotate -l foo=bar would break for
normal users if we changed to include by default.  The thing is, if we
include uninitialized in get, and someone is using that in pipe scripting
they would also be broken (i.e. kubectl get pods -o name | grep foo | xargs
kubectl annotate -f -).  So we wouldn't be backwards compatible (the reason
uninitialized are excluded is to keep them from impacting existing
workflows).

We have made the argument that kubectl get behaves differently with "-o"
(for show all) than without.

**Comment by smarterclayton** [comment]
I think the others are more clear.

**Comment by dixudx** [comment]
@caesarxuchao @ahmetb I would work on this based on #50445.

**Comment by deads2k** [comment]
> We plan to introduce --include-uninitialized as a global flag to kubectl, the default value depends on the subcommand, as discussed above.

A flag on our commands that defaults `false` seems like a good solution.  Clayton's points about all make a lot of sense.

**Comment by caesarxuchao** [comment]
The pipeline argument makes sense. How about:
* a `--include-uninitialized` global flag, default to `false`
* `kubectl delete <resource> --all`and `kubectl get <resource> --show-all`: include the uninitialized objects by default. Hide them if user explicitly set `--include-uninitialized=false`.
* `kubectl apply --prune`, similar to `kubectl delete <resource> --all`, will delete the uninitialized objects by default, unless user explicitly sets `--include-uninitialized=false`.
* `kubectl get <resource>`, `kubectl get <resouce> -w`: by default don't include uninitialized objects, unless user explicitly sets `--include-uninitialized=true`.
* `kubectl describe <resource>`: **includes uninitialized objects by default**. I think people don't use `describe in pipeline`?
* `kubectl set`, `kubectl edit`, `kubectl label`, `kubectl apply`, and `kubectl annotate`: by default don't include uninitialized objects, unless user explicitly sets `--include-uninitialized=true`.

**Comment by smarterclayton** [comment]
I think people using describe in pipeline are doing so for looser scripting (since we break that all the time).  I think include by default is ok.

Everything else I think I agree with.

**Comment by pwittrock** [comment]
@ahmetb Would you please update the issue description to contain the finalized result that we have reached consensus on - e.g. @caesarxuchao 's latest comment before this one

**Comment by dixudx** [comment]
@caesarxuchao @smarterclayton I've submitted a PR #50497 on this.

Some confusions here. Since the default value is `false`, what will make a difference by explicitly setting `--include-uninitialized=false`. That value will still be `false` unless we explicitly set it to `true`.

> a --include-uninitialized global flag, default to false

> unless user explicitly sets --include-uninitialized=false

**Comment by ahmetb** [comment]
@caesarxuchao Agree with the most of the proposed solution. One problem though:

I think `kubectl edit`, `kubectl apply` must include uninitialized objects. Imagine a resource like Deployment is failing to initialize. You would probably go `kubectl edit` to fix some annotation, or change it in the manifest and then `kubectl apply`, right? What would be the behavior of `kubectl apply` when an uninitialized Deployment object exists? In your proposal it seems to be creating a new object as if it never existed before.

**Comment by caesarxuchao** [comment]
> I think if you're referring to a resource by its name in a command, such as in edit/annotate/apply/label/set, you know it exists and you are trying to update it, so it must show.

Sorry for the confusion. My proposal is only about the batch mode of these commands, e.g., `kubectl edit pods`. I failed to copy the words "batch mode" to my last comment.

**Comment by caesarxuchao** [comment]
@dixudx can we have different default value of `--include-uninitialized` for different commands? Or can we detect if user explicitly sets a flag?

**Comment by ahmetb** [comment]
> My proposal is only about the batch mode of these commands

Got it, proposal LGTM.

**Comment by dixudx** [comment]
> Or can we detect if user explicitly sets a flag?

@caesarxuchao Yep, we can. Already revised PR #50497. PTAL. I think it's ready for review. Thanks.

Also for issue #49032 and #49035, @ahmetb they're somehow overlapped. Do we choose one of them? I've submitted PRs for them respectively. PTAL.

**Comment by ahmetb** [comment]
I started reviewing #50497, I need one more clarification:

> `kubectl set <resource>`, [...], kubectl label <resource>`, `kubectl annotate <resource>` [...] by default don't include uninitialized objects [...]

`kubectl set`/`annotate`/`label` only work with individual resources by referring to them by their name (or through -f) and don't work in batch mode. So I think they SHOULD include uninitialized objects by default and they shouldn't need to accept an `--include-uninitialized` flag as it's obsolete and excluding uninitialized doesn't make sense. Any objections?

**Comment by caesarxuchao** [comment]
@ahmetb I didn't run the command, but according to the kubectl help text:
```
  # Update all pods in the namespace
  kubectl annotate pods --all description='my frontend running nginx'
```

```
  # Update all deployments' and rc's nginx container's image to 'nginx:1.9.1'
  kubectl set image deployments,rc nginx=nginx:1.9.1 --all
```
```
  # Update all pods in the namespace
  kubectl label pods --all status=unhealthy
```

Such commands should have --include-unitialized=false by default, and allow users to set the flag to true.

**Comment by ahmetb** [comment]
@caesarxuchao thanks for correcting me. In this case, batch updates should NOT include uninitialized.

**Q1:** Do you think updates to individual resources i.e. `kubectl edit pods <name>` should include uninitialized or not? I think including it would be in line with the following:

> `kubectl describe <resource>`: **includes uninitialized objects by default**

-----

**Q2:** Can you also clarify does the proposal above work like the following?

> - `kubectl describe pods foo`: **includes uninitialized** by default
> - `kubectl get pods foo` by default **don't include** uninitialized objects

If so, it seems inconsistent to me. If the user explicitly creates an object with its name, I think both `get`/`describe` should include even though it is uninitialized. Having one of them work and the other one not sounds confusing. I'm having trouble finding a consistent framework explaining why this behavior is desirable.

**Comment by caesarxuchao** [comment]
@ahmetb 

> Do you think updates to individual resources i.e. kubectl edit pods <name> should include uninitialized or not?

Yes. Commands that spell out instance names (either via stdin or in the config file) should always include uninitialized objects. `--include-uninitialized` shouldn't affect such commands at all.

The entire proposal is about commands that use the LIST verb. The apiserver only hides uninitialized objecs from LIST operaions. Only `listOptions` has the `includeUninitialized` option.

> kubectl describe pods foo: includes uninitialized by default
kubectl get pods foo by default don't include uninitialized objects

No. Both should includes uninitialized objects. `--include-uninitialized` doesn't affect them at all.

**Comment by caesarxuchao** [comment]
@pwittrock @smarterclayton @ahmetb @dixudx in my original proposal I missed a detail.

Unlike `kubect get pods` or `kubectl edit pods`, `kubectl set/label/annotate` only operate in batch mode if `--all` or `-l` is specified. How about this:
* `kubectl set/label/annotate <resource> --all`, include uninitialized objects by default, unless user explicitly sets `--include-uninitiazlied=false`. This is to keep consistent with the semantic of `kubectl delete <resource> --all`. (I think Clayton agree with this: https://github.com/kubernetes/kubernetes/issues/49035#issuecomment-321125481)
* `kubectl set/label/annotate <resource> -l foo=bar` don't include uninitialized objects by default, unless user sets `--include-uninitiazlied=true`.

**Comment by dixudx** [comment]
ping @pwittrock @smarterclayton @ahmetb @caesarxuchao 
Updated. PTAL. Thanks.

**Comment by pwittrock** [comment]
The changes SGTM

**Comment by pwittrock** [comment]
@caesarxuchao Any updates?

**Comment by caesarxuchao** [comment]
PR was mostly lgtm yesterday, so we are on the right track.
